bson-codec-design/bson/document_decoder.go

```diff
@@ -0,0 +1,16 @@
+package bson
+
+// A DocumentDecoder transforms a Document into a Go type.
+//
+// A DocumentDecoder is goroutine safe and the DecodeDocument method can be called
+// from different goroutines simultaneously.
+type DocumentDecoder struct{}
```

Comment:
I'm not sure what value this provides over simply an Encoder/Decoder.

Reply:
It doesn't make sense for this to be the same type as the Encoder or Decoder unless those types are interfaces that are implemented by a concrete type that shares both. There is no reason why the `DocumentDecoder` or `DocumentEncoder` can't be goroutine safe, but the `Decoder` and `Encoder` types cannot be goroutine safe. This would mean we have to create additional `Decoder`s and `Encoder`s (for example in the mongo package on the Client), or use it incorrectly, or try to explain that certain methods on the `Decoder` and `Encoder` are goroutine safe. None of these seem pleasant.

Additionally, the `DocumentDecoder` and `DocumentEncoder` are doing something different from the `Decoder` and `Encoder` since they aren't really streaming, they are creating objects. The term "Decoder" and "Encoder" might not work best here, we could use something like `TransformTo` and `TransformFrom` but those are less pleasant.